domains/client/domain-executor/src/domain_bundle_proposer.rs

```diff
@@ -161,29 +162,36 @@ where
 
         let header_number = to_number_primitive(header_number);
 
-        // Ideally, the receipt of current block will be included in the next block, i.e., no
-        // missing receipts.
-        let receipts = if header_number == head_receipt_number + 1 {
-            let block_hash = self.client.hash(header_number.into())?.ok_or_else(|| {
-                sp_blockchain::Error::Backend(
-                    format!("Hash for Block {header_number:?} not found",),
-                )
-            })?;
-            vec![load_receipt(block_hash)?]
-        } else {
-            // Receipts for some previous blocks are missing.
-            let max_allowed = (head_receipt_number + max_drift).min(header_number);
-
-            let mut to_send = head_receipt_number + 1;
-            let mut receipts = Vec::with_capacity((max_allowed - to_send + 1) as usize);
-            while to_send <= max_allowed {
-                let block_hash = self.client.hash(to_send.into())?.ok_or_else(|| {
-                    sp_blockchain::Error::Backend(format!("Hash for Block {to_send:?} not found"))
+        let receipts = match header_number.cmp(&(head_receipt_number + 1)) {
+            // Ideally, the receipt of current block will be included in the next block, i.e., no
+            // missing receipts.
+            Ordering::Equal => {
+                let block_hash = self.client.hash(header_number.into())?.ok_or_else(|| {
+                    sp_blockchain::Error::Backend(format!(
+                        "Hash for Block {header_number:?} not found",
+                    ))
                 })?;
-                receipts.push(load_receipt(block_hash)?);
-                to_send += 1;
+                vec![load_receipt(block_hash)?]
+            }
+            // Receipts for some previous blocks are missing.
+            Ordering::Greater => {
+                let max_allowed = (head_receipt_number + max_drift).min(header_number);
+                let mut to_send = head_receipt_number + 1;
+                let mut receipts = Vec::with_capacity((max_allowed + 1 - to_send) as usize);
+                while to_send <= max_allowed {
+                    let block_hash = self.client.hash(to_send.into())?.ok_or_else(|| {
+                        sp_blockchain::Error::Backend(format!(
+                            "Hash for Block {to_send:?} not found"
+                        ))
+                    })?;
+                    receipts.push(load_receipt(block_hash)?);
+                    to_send += 1;
+                }
+                receipts
             }
-            receipts
+            // `header_number` is less or equal to `head_receipt_number` means the primary chain
+            // is ahead of the execution chain thus there is no receipt to submit.
+            Ordering::Less => vec![],
```

Comment:
Returning empty receipts directly does not make a lot of sense and it contradicts this runtime check:

https://github.com/subspace/subspace/blob/c7f738a59213271d55db300aaa5a91d2b3136716/crates/pallet-domains/src/lib.rs#L542-L546

When the executor finds the domain is even lagging behind the receipt chain on the primary chain, it should discard this bundle and catch up with the primary chain first. In addition, I think we should check whether it's _ready_ before attempting to produce a new bundle, if it's even behind the best receipt, then just skip this slot.
https://github.com/subspace/subspace/blob/c7f738a59213271d55db300aaa5a91d2b3136716/domains/client/domain-executor/src/domain_worker.rs#L194

Reply:
A check is added to skip the slot if the executor is behind the best receipt, the previous changes are also preserved to make it explicit.